src/navigators/createMaterialTopTabNavigator.js

```diff
@@ -113,18 +133,109 @@ class TabView extends React.PureComponent<Props> {
 
   _renderPanPager = props => <TabViewPagerPan {...props} />;
 
-  _renderScene = ({ route, focused }) => {
-    const { renderScene, animationEnabled, swipeEnabled } = this.props;
+  componentWillMount() {
+    const { animationEnabled } = this.props;
 
-    if (animationEnabled === false && swipeEnabled === false) {
-      return (
-        <ResourceSavingScene isFocused={focused}>
-          {renderScene({ route })}
-        </ResourceSavingScene>
+    // We rely on action listener to animate from the current index to the next one, no need to listen without animation
+    if (animationEnabled) {
+      this._actionListener = this.props.navigation.addListener(
+        'action',
+        this._onAction
       );
     }
+  }
+
+  componentWillUnmount() {
+    if (this._actionListener) {
+      this._actionListener.remove();
+    }
+
+    if (this._transitionTimeout) {
+      clearTimeout(this._transitionTimeout);
+    }
+  }
+
+  componentWillReceiveProps(nextProps) {
+    if (
+      nextProps.navigation.state.index !== this.props.navigation.state.index
+    ) {
+      const { index } = nextProps.navigation.state;
+
+      this.setState(state => ({
+        loaded: state.loaded.includes(index)
+          ? state.loaded
+          : [...state.loaded, index],
+      }));
+    }
+  }
+
+  _onAction = payload => {
```

Comment:
We shouldn't need to subscribe to navigation actions. We already get state updates via the `navigation.state` prop. We should react to it in `componentDidUpdate` or use the `onIndexChange` callback if we want to react to it before the update happens. This will remove the need to maintain a subscription which is cumbersome.